dot/sync/chain_sync.go

```diff
@@ -609,19 +678,41 @@ func (cs *chainSync) doSync(req *network.BlockRequestMessage) *workerError {
 	// response was validated! place into ready block queue
 	for _, bd := range resp.BlockData {
 		// block is ready to be processed!
-		logger.Trace("new ready block", "hash", bd.Hash, "number", bd.Header.Number)
-		cs.pendingBlocks.removeBlock(bd.Hash)
-		cs.readyBlocks <- bd
+		handleReadyBlock(bd, cs.pendingBlocks, cs.readyBlocks)
 	}
 
 	return nil
 }
 
+func handleReadyBlock(bd *types.BlockData, pendingBlocks DisjointBlockSet, readyBlocks *blockQueue) {
+	// see if there are any descendents in the pending queue that are now ready to be processed,
+	// as we have just become aware of their parent block
+
+	// if header was not requested, get it from the pending set
+	// if we're expecting headers, validate should ensure we have a header
+	if bd.Header == nil {
+		block := pendingBlocks.getBlock(bd.Hash)
+		if block.header == nil {
+			panic("block isn't ready! header is unknown!")
```

Comment:
any way to avoid having panics?

Reply:
yeah this shouldn't happen ever, i can remove it